### DISCUSSION TITLE
How to retrieve refresh token from Google?

### DISCUSSION THREAD

**Discussion by aaaguirrep** [panva/openid-client#347] [Q&A]
Hi, I am using openid-client library in my web application, I am authenticating the google users without issues but I can't retrieve refresh_token, I have tested the next code but I can't see refresh_token. I am using google as IDP.

```javascript
    const login_url = client.authorizationUrl({
        scope: 'openid email profile',
        code_challenge,
        code_challenge_method: 'S256'
        access_type: 'offline'
    })
```

I have checked the documentation from google and it specifies to pass access_type to offline.
https://developers.google.com/identity/protocols/oauth2/openid-connect#refresh-tokens

Can you help me?

Thanks in advance.

Regards,
Arsenio

**Comment by panva** [comment]
Literally taken from the google docs you linked.

> Refresh tokens
> In your request for API access you can request a refresh token to be returned during the code exchange. A refresh token provides your app continuous access to Google APIs while the user is not present in your application. To request a refresh token, add set the access_type parameter to offline in your authentication request.
> 
> Considerations:
> 
> - Be sure to store the refresh token safely and permanently, **because you can only obtain a refresh token the first time that you perform the code exchange flow**.
>
> - **There are limits on the number of refresh tokens that are issued: one limit per client/user combination, and another per user across all clients**. If your application requests too many refresh tokens, it may run into these limits, in which case older refresh tokens stop working.

**Reply by aaaguirrep** [reply]
Thanks @panva for your help and congratulation for the library. I have retrieved the refresh_token, I did a new authentication process and it worked. I was not trying with the first login.
